tests/collections/llm/llama3_pretraining.py

```diff
@@ -0,0 +1,130 @@
+        '--precision', type=str, choices=['bf16', 'fp16', 'fp32'], default='bf16', help="Override recipe precision"
+    )
+    parser.add_argument('--fp8', action='store_true', help="Enable FP8")
+
+    return parser.parse_args()
+
+
+def main():
+    args = get_args()
+
+    exp_name = "L2_llama3_small_pretrain_test"
+    pretrain_recipe = llm.llama3_8b.pretrain_recipe(
+        dir=args.experiment_dir, name=exp_name, num_gpus_per_node=args.devices
+    )
+
+    pretrain_recipe.model = llm.LlamaModel(small_llama_cfg(args.seq_length))
+
+    if args.data_path and args.tokenizer_path:
+        pretrain_recipe.data = train_data(
+            data_path=args.data_path,
+            tokenizer_path=args.tokenizer_path,
+            index_mapping_dir=args.index_mapping_dir,
+            seq_length=args.seq_length,
+        )
+
+    # Recipe Overrides
+    pretrain_recipe.trainer.max_steps = args.max_steps
+    pretrain_recipe.trainer.log_every_n_steps = 1
+    pretrain_recipe.log.ckpt.every_n_train_steps = None
+    pretrain_recipe.trainer.val_check_interval = 2
+
+    if args.early_stop:
+        pretrain_recipe.trainer.callbacks.append(StopBeforeEnd(stop_on_step=args.early_stop))
+
+    if not args.precision == 'bf16' or args.fp8:  # default case is bf16 without fp8
+        import llm.recipes.precision.mixed_precision as mp_recipes
+
+        key = (args.precision, args.fp8)
+        precision_recipe = {
+            ("fp16", False): mp_recipes.fp16_mixed,
+            ("bf16", True): mp_recipes.bf16_with_fp8_mixed,
+            ("fp16", True): mp_recipes.fp16_with_fp8_mixed,
+            # Need fp32
+        }[key]
+        pretrain_recipe.trainer.plugins = precision_recipe()
+    dtype_map = {"bf16": torch.bfloat16, "fp16": torch.float16, "fp32": torch.float32}
+    debugger_callback = ParameterDebugger(
+        param_fn=create_verify_precision(dtype_map[args.precision]),
+        grad_fn=create_verify_precision(torch.float32),
+        log_on_hooks=["on_train_start", "on_train_end"],
+    )
+    pretrain_recipe.trainer.callbacks.append(debugger_callback)
+
+    parallelisms = {
+        "tensor_model_parallel_size": args.tp,
+        "pipeline_model_parallel_size": args.pp,
+        "virtual_pipeline_model_parallel_size": args.vp,
+        "context_parallel_size": args.cp,
+        "sequence_parallel": bool(args.sp) if args.sp is not None else None,
+    }
+    for k, v in parallelisms.items():
+        if v is not None:  # use recipe default if not specified
+            setattr(pretrain_recipe.trainer.strategy, k, v)
+        parallelisms[k] = getattr(pretrain_recipe.trainer.strategy, k)
+    pretrain_recipe.trainer.callbacks.append(MCoreModelAttributeValidator(parallelisms))
+
+    run.run(pretrain_recipe, direct=True)
+
+    verify_ckpt_dir(
+        pretrain_recipe.log.ckpt,
+        args.early_stop or args.max_steps,
+        pretrain_recipe.trainer.val_check_interval,
+        os.path.join(args.experiment_dir, exp_name),
+    )
+
```

Comment:
Is it possible to call the `generate` function here to test inference on the generated checkpoint as well?

Reply:
this is a great idea. i can work on adding this in a separate PR.

Reply:
i don't know if the checkpoint would need any conversion, do you know?